xbmc/pvr/channels/PVRChannelGroup.cpp

```diff
@@ -697,43 +722,59 @@ bool CPVRChannelGroup::Renumber(void)
 
   bool bReturn(false);
   unsigned int iChannelNumber(0);
-  bool bUseBackendChannelNumbers(CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERS) &&
+  bool bUsingBackendChannelNumbers(CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_PVRMANAGER_USEBACKENDCHANNELNUMBERS) &&
                                  CServiceBroker::GetPVRManager().Clients()->EnabledClientAmount() == 1);
+  bool bStartGroupChannelNumbersFromOne(CServiceBroker::GetSettingsComponent()->GetSettings()->GetBool(CSettings::SETTING_PVRMANAGER_STARTGROUPCHANNELNUMBERSFROMONE) &&
+                                        !bUsingBackendChannelNumbers);
 
   CSingleLock lock(m_critSection);
 
   CPVRChannelNumber currentChannelNumber;
-  for (PVR_CHANNEL_GROUP_SORTED_MEMBERS::iterator it = m_sortedMembers.begin(); it != m_sortedMembers.end(); ++it)
+  CPVRChannelNumber currentClientChannelNumber;
+  for (auto& sortedMember : m_sortedMembers)
   {
-    if ((*it).channel->IsHidden())
+    currentClientChannelNumber = sortedMember.clientChannelNumber;
+
+    if (sortedMember.channel->IsHidden())
     {
       currentChannelNumber = CPVRChannelNumber(0, 0);
     }
-    else if (bUseBackendChannelNumbers)
-    {
-      currentChannelNumber = (*it).channel->ClientChannelNumber();
-    }
     else
     {
       if (IsInternalGroup())
+      {
         currentChannelNumber = CPVRChannelNumber(++iChannelNumber, 0);
+      }
       else
-        currentChannelNumber = m_allChannelsGroup->GetChannelNumber((*it).channel);
+      {
+        if (bStartGroupChannelNumbersFromOne)
+          currentChannelNumber = CPVRChannelNumber(++iChannelNumber, 0);
+        else
+          currentChannelNumber = m_allChannelsGroup->GetChannelNumber(sortedMember.channel);
+
+        if (!sortedMember.clientChannelNumber.IsValid())
+          currentClientChannelNumber = m_allChannelsGroup->GetClientChannelNumber(sortedMember.channel);
+      }
     }
 
-    if ((*it).channelNumber != currentChannelNumber)
+    if (sortedMember.channelNumber != currentChannelNumber || sortedMember.clientChannelNumber != currentClientChannelNumber)
     {
       bReturn = true;
       m_bChanged = true;
-      (*it).channelNumber = currentChannelNumber;
-    }
+      sortedMember.channelNumber = currentChannelNumber;
+      sortedMember.clientChannelNumber = currentClientChannelNumber;
 
-    //! @todo This is a quick fix for v18. Whole channel number handling should be reworked - code is imo unmaintainable.
-    if (IsInternalGroup())
-      (*it).channel->SetChannelNumber((*it).channelNumber);
```

Comment:
Because each channel now uses its group specific channel number as the source of this value. The channel number on the channel itself is only used to display on the UI and gets set per selected group.

Reply:
Okay.